src/com/amazon/ionelement/api/Equivalence.kt

```diff
@@ -0,0 +1,127 @@
+    if (annotations != other.annotations) return false
+    // Metas intentionally not included here.
+
+    return if (isNull) {
+        // Already verified that they are the same type
+        other.isNull
+    } else if (other.isNull) {
+        false
+    } else
+    // Matching an enum rather than a type allows the Kotlin compiler
+    // to use a table switch instead of a chain of if/else comparisons.
+        when (thisType) {
+            BOOL -> booleanValue == other.booleanValue
+            INT -> when {
+                integerSize != other.integerSize -> false
+                integerSize == IntElementSize.LONG -> longValue == other.longValue
+                else -> bigIntegerValue == other.bigIntegerValue
+            }
+            // compareTo() distinguishes between 0.0 and -0.0 while `==` operator does not.
+            FLOAT -> doubleValue.compareTo(other.doubleValue) == 0
+            // `==` considers `0d0` and `-0d0` to be equivalent.  `Decimal.equals` does not.
+            DECIMAL -> Decimal.equals(decimalValue, other.decimalValue)
+            TIMESTAMP -> timestampValue == other.timestampValue
+            STRING -> stringValue == other.stringValue
+            SYMBOL -> symbolValue == other.symbolValue
+            BLOB -> blobValue == other.blobValue
+            CLOB -> clobValue == other.clobValue
+            LIST -> listValues == other.listValues
+            SEXP -> sexpValues == other.sexpValues
+            STRUCT -> {
+                val thisFields = this.structFields
+                val otherFields = other.structFields
+                when {
+                    thisFields === otherFields -> true
+                    thisFields.size != otherFields.size -> false
+                    // We've tried the inexpensive checks, now do a deep comparison
+                    else -> thisFields.groupingBy { it }.eachCount() == otherFields.groupingBy { it }.eachCount()
+                }
+            }
+            NULL -> TODO("Unreachable")
+        }
+}
+
+/**
+ * Calculates the hash code of an [IonElement].
+ *
+ * Implementations of [IonElement] MAY NOT calculate their own hash codes, but MAY cache the result of this function.
+ */
+public fun hashElement(ionElement: IonElement): Int {
+    val element = ionElement.asAnyElement()
+    val typeAndValueHashCode = if (element.isNull) {
+        element.type.hashCode()
+    } else {
+        // Matching an enum rather than a type allows the Kotlin compiler
+        // to use a tableswitch instead of a chain of if/else comparisons.
+        val valueHashCode = when (element.type) {
+            BOOL -> element.booleanValue.hashCode()
+            INT -> when (element.integerSize) {
+                IntElementSize.LONG -> element.longValue.hashCode()
+                IntElementSize.BIG_INTEGER -> element.bigIntegerValue.hashCode()
+            }
+            // Adding compareTo(0.0) causes 0e0 to have a different hash code than -0e0
```

Comment:
... which isn't _incorrect_ though

Reply:
Right—it's not incorrect for them to have the same hashcode. This is just explaining why we are using `compareTo`. If this ends up being a problem, we can change it.